performance/test-perf.js

```diff
@@ -0,0 +1,405 @@
+				Math.floor((result.ticks / threshold) * 100) + "% of threshold" +
+			")"
+		);
+
+		o(result.ticks <= threshold).equals(true)(result.ticks + " ticks, " + threshold + " allowed");
+
+		done();
+	};
+}
+
+function loop(iter, time) {
+	var start = now(),
+		count = 0;
+
+	// Run as many instances of iter as possible
+	while (now() - start < time) {
+		count++;
+		iter();
+	}
+
+	return count;
+}
+
+function benchmark(iter, callback) {
+	var a = 0, // eslint-disable-line no-unused-vars
+		count = 5,
+		time = 500,
+		passes = 0,
+		total = 0,
+		noops, i;
+
+	function noop() {
+		try { a++; } finally { a += Math.random(); }
+	}
+
+	// warm
+	for(i=100; i--;) {
+		noop();
+		iter();
+	}
+
+	// Count how many noops() occur in the time period
+	noops = loop(noop, time);
+
+	// run iter() through loop() `count` times
+	function next() {
+		total += loop(iter, time);
+
+		setTimeout(++passes === count ? done : next, 10);
+	}
+
+	function done() {
+		callback({
+			total : total,
+			noops : noops,
+			count : count,
+			time  : time,
+			ticks : Math.round(noops / total * count),
+			hz    : total / count / time * 1000
+		});
+	}
+
+	next();
+}
+
+o.spec("perf", function() {
+	var m, scratch
+
+	o.before(function () {
+		var doc = typeof document !== "undefined" ? document : null;
+
+		if(!doc) {
+			var mock = browserMock()
+			if (typeof global !== "undefined") { global.window = mock }
+
+			doc = mock.document;
+		}
+
+		m = require("../mithril") // eslint-disable-line global-require
+
+		scratch = doc.createElement("div");
+		(doc.body || doc.documentElement).appendChild(scratch);
+	})
+
+	o.afterEach(function () {
+		scratch.innerHTML = "";
+	})
+
+	o.after(function () {
+		scratch = null
+	})
+
+	o("rerender without changes", function (done, timeout) {
+		timeout(5000);
+
+		var vdom = m("div", {"class": "foo bar", "data-foo": "bar", p: 2},
+			m("header",
+				m("h1", {"class": "asdf"}, "a ", "b", " c ", 0, " d"),
+				m("nav",
+					m("a", {href: "/foo"}, "Foo"),
+					m("a", {href: "/bar"}, "Bar")
+				)
+			),
+			m("main",
+				m("form", {onSubmit: function onSubmit() {}},
+					m("input", {type: "checkbox", checked: true}),
+					m("input", {type: "checkbox", checked: false}),
+					m("fieldset",
+						m("label",
+							m("input", {type: "radio", checked: true})
+						),
+						m("label",
+							m("input", {type: "radio"})
+						)
+					),
+					m("button-bar",
+						m("button",
+							{style: "width:10px; height:10px; border:1px solid #FFF;"},
+							"Normal CSS"
+						),
+						m("button",
+							{style: "top:0 ; right: 20"},
+							"Poor CSS"
+						),
+						m("button",
+							{style: "invalid-prop:1;padding:1px;font:12px/1.1 arial,sans-serif;", icon: true},
+							"Poorer CSS"
+						),
+						m("button",
+							{style: {margin: 0, padding: "10px", overflow: "visible"}},
+							"Object CSS"
+						)
+					)
+				)
+			)
+		);
+
+		benchmark(
+			function () {
+				m.render(scratch, vdom);
```

Comment:
... but then, I think we have tests that verify that the update phase doesn't run on cached nodes... well, I guess some redundance can't hurt.